src/Introspecter.js

```diff
@@ -0,0 +1,104 @@
+import React, {useContext, useEffect, useState} from 'react';
+import PropTypes from 'prop-types';
+import {Loading, useDataProvider} from 'react-admin';
+import ResourceSchemaAnalyzerContext from './ResourceSchemaAnalyzerContext';
+
+const IntrospecterComponent = ({
+  component: Component,
+  resourceSchemaAnalyzer,
+  includeDeprecated,
+
+  resource,
+  resources,
+  fetching,
+  error,
+
+  ...rest
+}) => {
+  if (fetching) {
+    return <Loading />;
+  }
+
+  if (error) {
+    if ('production' === process.env.NODE_ENV) {
+      console.error(error);
+    }
+
+    throw new Error('API schema is not readable');
+  }
+
+  if (resources == null) {
+    return null;
+  }
+
+  const resourceSchema = resources.find(r => r.name === resource);
+
+  if (!resourceSchema || !resourceSchema.fields) {
+    if ('production' === process.env.NODE_ENV) {
+      console.error(`Resource ${resource} not present inside API description`);
+    }
+
+    throw new Error(`Resource ${resource} not present inside API description`);
+  }
+
+  const fields = includeDeprecated
+    ? resourceSchema.fields
+    : resourceSchema.fields.filter(({deprecated}) => !deprecated);
+
+  return (
+    <Component
+      resourceSchemaAnalyzer={resourceSchemaAnalyzer}
+      resource={resource}
+      resourceSchema={resourceSchema}
+      fields={fields}
+      {...rest}
+    />
+  );
+};
+
+const Introspecter = ({
+  component,
+  includeDeprecated = false,
+  resource,
+  ...rest
+}) => {
+  const dataProvider = useDataProvider();
+  const resourceSchemaAnalyzer = useContext(ResourceSchemaAnalyzerContext);
+  const [resources, setResources] = useState();
+  const [fetching, setFetching] = useState(true);
+  const [error, setError] = useState();
+
+  useEffect(() => {
+    dataProvider
+      .introspect()
```

Comment:
Am I understanding correctly: as `Introspecter` is called in every `Guesser`, this will fetch the entire hydra schema once for each resource, field, input? If so, that's a huge waste of bandwidth and a major slowdown. Especially since you've already fetched the resources in AdminGuesser.

Reply:
The introspection logic is only called once then fetched from a local cache: https://github.com/api-platform/admin/blob/b022e157db00378df57cf22e544bf33ba55dd173/src/hydra/dataProvider.js#L116